fw/http_parser.c

```diff
@@ -3320,6 +3332,167 @@ __parse_pragma(TfwHttpMsg *hm, unsigned char *data, size_t len)
+	 */
+	__FSM_STATE(I_UpgradeProtocol) {
+		__FSM_I_MATCH_MOVE_fixup(token, I_UpgradeProtocol,
+					 TFW_STR_NAME);
+		__FSM_I_JMP(I_UpgradeProtocolEnd);
+	}
+
+	__FSM_STATE(I_UpgradeProtocolEnd) {
+		if (likely(__fsm_sz)) {
+			/* Save protocol name */
+			__msg_hdr_chunk_fixup(p, __fsm_sz);
+			__FSM_I_chunk_flags(TFW_STR_NAME);
+		}
+
+#define TFW_STR_PROTOCOL_WEBSOCKET "websocket"
+		if (!test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET,
+				msg->flags))
+		{
+			/* Current header 'upgrade:' not always lengthy than
+			 * 'websocket' protocol name length being matched.
+			 * Need to compare lengths before matching.
+			 */
+			if ((msg->stream->parser.hdr.len >=
+				SLEN(TFW_STR_PROTOCOL_WEBSOCKET))
+			    && tfw_str_eq_cstr_off(
+				&msg->stream->parser.hdr,
+				msg->stream->parser.hdr.len -
+					SLEN(TFW_STR_PROTOCOL_WEBSOCKET),
+				TFW_STR_PROTOCOL_WEBSOCKET,
+				SLEN(TFW_STR_PROTOCOL_WEBSOCKET),
+				TFW_STR_EQ_CASEI))
+			{
+				__set_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET,
+						msg->flags);
+			}
+		}
+#undef TFW_STR_PROTOCOL_WEBSOCKET
+
+		if (likely(IS_CRLF(*(p + __fsm_sz)))) {
+			__FSM_EXIT(__data_processed(p + __fsm_sz));
+		}
+
+		p += __fsm_sz;
+		switch (*p) {
+		case '/':
+			__FSM_I_MOVE_fixup(I_UpgradeVersionStart, 1, 0);
+		case ',':
+			__FSM_I_JMP(I_UpgradeComma);
+		default:
+			return CSTR_NEQ;
+		}
+	}
+
+	/*
+	 * Protocol version stored in a separate value TfwStr chunk.
+	 * May not be empty. '/' already matched.
+	 */
+	__FSM_STATE(I_UpgradeVersionStart) {
+		__FSM_I_MATCH_MOVE_fixup(token, I_UpgradeVersion,
+					 TFW_STR_VALUE);
+		if (likely(__fsm_sz))
+			__FSM_I_JMP(I_UpgradeVersionEnd);
+		return CSTR_NEQ;
+	}
+
+	/*
+	 * At this state we know that we saw at least one character in
+	 * protocol version and now we can pass zero length token.
+	 */
+
+	__FSM_STATE(I_UpgradeVersion) {
+		__FSM_I_MATCH_MOVE_fixup(token, I_UpgradeVersion,
+					 TFW_STR_VALUE);
+		__FSM_I_JMP(I_UpgradeVersionEnd);
+	}
+
+	__FSM_STATE(I_UpgradeVersionEnd) {
+		if (likely(__fsm_sz)) {
+			/* Save protocol version */
+			__msg_hdr_chunk_fixup(p, __fsm_sz);
+			__FSM_I_chunk_flags(TFW_STR_VALUE);
+		}
+
+		if (likely(IS_CRLF(*(p + __fsm_sz)))) {
+			__FSM_EXIT(__data_processed(p + __fsm_sz));
+		}
+
+		p += __fsm_sz;
+		if (likely(*p == ','))
+			__FSM_I_JMP(I_UpgradeComma);
+		return CSTR_NEQ;
+	}
+
+	/* ',' was already matched */
+	__FSM_STATE(I_UpgradeComma) {
+		/*
+		 * Fixup current delimiters chunk and move to next protocol
+		 * if we can eat ',' and SP at once.
+		 */
+		if (likely(__data_available(p, 2))) {
+			if (likely(*(p + 1) == ' '))
+				__FSM_I_MOVE_fixup(I_UpgradeProtocolStart, 2,
+						   0);
+			return CSTR_NEQ;
+		}
+
+		/*
+		 * Only ',' is available now: fixup ',' as independent chunk,
+		 * SP will be fixed up at next enter to the FSM.
+		 */
+		__FSM_I_MOVE_fixup(I_UpgradeSP, 1, 0);
+	}
+
+	__FSM_STATE(I_UpgradeSP) {
+		if (unlikely(*p != ' '))
+			return CSTR_NEQ;
```

Comment:
The RFC doesn't allow many SPs, right? Could you please leave comment if so

Reply:
In fact this part of patch inconsistent with the RFC. Lists have to be matched a little different. I'll fix that.